recipes/openblas/all/conanfile.py

```diff
@@ -19,134 +23,219 @@ class OpenblasConan(ConanFile):
         "fPIC": [True, False],
         "build_lapack": [True, False],
         "use_thread": [True, False],
+        "use_openmp": [True, False],
         "dynamic_arch": [True, False],
+        "with_avx512": [True, False],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
         "build_lapack": False,
         "use_thread": True,
+        "use_openmp": False,
         "dynamic_arch": False,
+        "with_avx512": False,
     }
-    generators = "cmake"
+    package_type = "library"
     short_paths = True
 
+    def _fortran_runtime(self, fortran_id):
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            if fortran_id == "GNU":
+                if self.settings.compiler == "gcc":
+                    # Compiler vs. gfortran runtime ver.: 5,6: 3, 7: 4, >=8: 5
+                    if Version(self.settings.compiler.version).major >= "5":
+                        return "gfortran"
+                if self.settings.compiler == "clang":
+                    if Version(self.settings.compiler.version).major > "8":
+                        return "gfortran"  # Runtime version gfortran5
+
+        self.output.warning(
+            f"Unable to select runtime for Fortran {fortran_id} "
+            f"and C++ {self.settings.compiler} {self.settings.compiler.version}")
+        return None
+
     @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+    def _openmp_runtime(self):
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            if self.settings.compiler == "gcc":
+                if Version(self.settings.compiler.version).major > "7":
+                    return "gomp"
+            if self.settings.compiler == "clang":
+                if Version(self.settings.compiler.version).major > "8":
+                    return "omp"
+        return None
 
     @property
-    def _build_subfolder(self):
-        return "build_subfolder"
+    def _fortran_compiler(self):
+        comp_exe = self.conf.get("tools.build:compiler_executables")
+        if comp_exe and 'fortran' in comp_exe:
+            return comp_exe["fortran"]
+        return environ.get('FC')
 
     def export_sources(self):
-        self.copy("CMakeLists.txt")
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
+        # Build LAPACK by default if possible w/o Fortran compiler
+        if Version(self.version) >= "0.3.21":
+            self.options.build_lapack = True
+
     def configure(self):
         if self.options.shared:
-            del self.options.fPIC
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def package_id(self):
+        if self.info.options.build_lapack:
+            # Capture fortran compiler in package id (if found)
+            f_compiler = self._fortran_compiler
+            if not f_compiler:
```

Comment:
Being overly strict with the fortran compiler version added fully to the ``package_id`` will easily result in "missing binary" for all users, all the time, because of minimal variations in their locally installed ``gfortran`` version, that might not even be used for building. So probably we don't want to auto detect the version here with ``compiler --version``, and if that is the case, it should be taken from the declarative Conan ``settings`` or ``conf``

Reply:
With my latest suggestion, the `package_id` option `build_lapack` is changed to include Fortran compiler id, as well as major and minor version. This is done using a CMake helper script, which is run in package_id(), with its limitations at version 3.15 (not all Fortran compiler ids are detected correctly, e.g. Classic Flang or Intel LLVM). Generally, it is laborious to properly map Fortran ABI compatibility into the `package_id`, but at least removing `patch` might reduce the occurrence of "missing binary". Ideally CMake >=3.24 should be used, but I believe build_requirements are evaluated later and not available in `package_id()`.

Since OpenBLAS 0.3.21, Fortran is not required. In this case, if the user do not specify `fortran` in `tools.build:compiler_executable`, it is built without Fortran and therefore the `package_id` is not updated with any Fortran information.

Note that this build option is not enabled on CCI (except perhaps since 0.3.23).

I admit that this way of doing it is rather cumbersome and are open for better suggestions. I have tried to avoid using a custom `settings.yml`, because of the simplicity of not needing to distribute custom `settings.yml` files.